plugin/pulpcore/plugin/stages/artifact_stages.py

```diff
@@ -289,15 +289,17 @@ class ArtifactSaver(Stage):
             The coroutine for this stage.
         """
         async for batch in self.batches(in_q):
-            artifacts_to_save = []
+            da_to_save = []
             for declarative_content in batch:
                 for declarative_artifact in declarative_content.d_artifacts:
                     if declarative_artifact.artifact.pk is None:
                         declarative_artifact.artifact.file = str(declarative_artifact.artifact.file)
-                        artifacts_to_save.append(declarative_artifact.artifact)
+                        da_to_save.append(declarative_artifact)
 
-            if artifacts_to_save:
-                Artifact.objects.bulk_get_or_create(artifacts_to_save)
+            if da_to_save:
+                for da, artifact in zip(da_to_save, Artifact.objects.bulk_get_or_create(
```

Comment:
Are we sure, that the order is preserved here?

Reply:
Yeah this has been a key question we've been asking. It is not documented in their docs as something we can rely on. So that makes me a little nervous also. @dralley and I were talkinga bout this and he reviewed the bulk_create implementation to ensure it is order-preserving. We also checked that our implementation of bulk_get_or_create is also order-preserving.

We don't have a good way of guaranteeing this from django going forward. Maybe we should have a unit test so we can assert bulk_create in latest always returns items in-order. That would be one way to raise our confidence.

Reply:
OK, let's hope for the best.
But i still think, we rely heavily on bulk_create to return the pk's here.
And if i read https://docs.djangoproject.com/en/2.1/ref/models/querysets/#bulk-create correctly, this is not a given on anything but postgres.

Reply:
@mdellweg this is also concerning to me. We need to have a convo on the mailing list about if this is acceptable. This issue was introduced when we switched from UUID to ints since the uuid's were client side generated.